collector/lib/CollectorConfig.cpp

```diff
@@ -422,33 +425,106 @@ void CollectorConfig::YamlConfigToConfig(YAML::Node& yamlConfig) {
+  CLOG(INFO) << "Runtime configuration:\n" + GetRuntimeConfigStr();
 
   return;
 }
 
-void CollectorConfig::HandleConfig(const std::filesystem::path& filePath) {
+bool CollectorConfig::HandleConfig(const std::filesystem::path& filePath) {
   if (!std::filesystem::exists(filePath)) {
     CLOG(DEBUG) << "No configuration file found. " << filePath;
-    return;
+    return true;
   }
 
   try {
     YAML::Node yamlConfig = YAML::LoadFile(filePath);
     YamlConfigToConfig(yamlConfig);
   } catch (const YAML::BadFile& e) {
-    CLOG(FATAL) << "Failed to open the configuration file: " << filePath
-                << ". Error: " << e.what();
+    CLOG(ERROR) << "Failed to open the configuration file: " << filePath << ". Error: " << e.what();
+    return false;
   } catch (const YAML::ParserException& e) {
-    CLOG(FATAL) << "Failed to parse the configuration file: " << filePath
-                << ". Error: " << e.what();
+    CLOG(ERROR) << "Failed to parse the configuration file: " << filePath << ". Error: " << e.what();
+    return false;
   } catch (const YAML::Exception& e) {
-    CLOG(FATAL) << "An error occurred while loading the configuration file: " << filePath
-                << ". Error: " << e.what();
+    CLOG(ERROR) << "An error occurred while loading the configuration file: " << filePath << ". Error: " << e.what();
+    return false;
   } catch (const std::exception& e) {
-    CLOG(FATAL) << "An unexpected error occurred while trying to read: " << filePath << e.what();
+    CLOG(ERROR) << "An unexpected error occurred while trying to read: " << filePath << e.what();
+    return false;
+  }
+
+  return true;
+}
+
+void CollectorConfig::WaitForFileToExist(const std::filesystem::path& filePath) {
+  while (!std::filesystem::exists(filePath)) {
+    if (runtime_config_.has_value()) {
+      CLOG(INFO) << "The configuration file has been deleted. Resetting the configuration.";
+      runtime_config_.reset();
+    }
+    sleep(1);
+  }
+}
+
+int WaitForInotifyAddWatch(int fd, const std::filesystem::path& filePath) {
+  while (true) {
+    int wd = inotify_add_watch(fd, filePath.c_str(), IN_MODIFY | IN_MOVE_SELF | IN_DELETE_SELF);
+    if (wd < 0) {
+      CLOG_THROTTLED(ERROR, std::chrono::seconds(30)) << "Failed to add inotify watch for " << filePath;
+      close(fd);
+    } else {
+      return wd;
+    }
+    sleep(1);
+  }
+}
+
+void CollectorConfig::WatchConfigFile(const std::filesystem::path& filePath) {
+  int fd = inotify_init();
+  if (fd < 0) {
+    CLOG(ERROR) << "inotify_init() failed: " << StrError();
+    CLOG(ERROR) << "Runtime configuration will not be used.";
+    return;
+  }
+  WaitForFileToExist(filePath);
+  int wd = WaitForInotifyAddWatch(fd, filePath);
+  bool success = HandleConfig(filePath);
+  if (!success) {
+    CLOG(FATAL) << "Unable to parse configuration file: " << filePath;
   }
+
+  char buffer[1024];
+  while (true) {
+    int length = read(fd, buffer, sizeof(buffer));
```

Comment:
C++ understanding: Is my understanding correct that `inotify` registers a file descripter on `fd`, after registering an infinite loop is started to `read` the fd for the inotify events into the `buffer`.

The buffer is casted to an `inotify_event` and we check if it is the `IN_MODIFY` event.

Basically, how I understand this now, is that the we need to handle the inotify events and wait for `read` continuously,.

Is `read` a blocking call or would the loop be halted until an event is tirggered?

Reply:
> C++ understanding: Is my understanding correct that `inotify` registers a file descripter on `fd`, after registering an infinite loop is started to `read` the fd for the inotify events into the `buffer`.
> 
> The buffer is casted to an `inotify_event` and we check if it is the `IN_MODIFY` event.
> 
> Basically, how I understand this now, is that the we need to handle the inotify events and wait for `read` continuously,.

Yes, this is all correct.

> Is `read` a blocking call or would the loop be halted until an event is tirggered?

In this instance, `read` is a blocking call, so the thread will stop until an event occurs, from `man inotify`:
```
   Reading events from an inotify file descriptor
       To determine what events have occurred, an application read(2)s from the inotify file descriptor.  If no events have
       so far occurred, then, assuming a blocking file descriptor, read(2) will block until at least one event occurs  (un‐
       less interrupted by a signal, in which case the call fails with the error EINTR; see signal(7)).
```

Reply:
Thanks Mauro!